client/web/src/enterprise/batches/create/workspaces-preview/WorkspacesPreview.tsx

```diff
@@ -1,174 +1,206 @@
+    // have to do this outside of Apollo Client because we continue to requery the
+    // workspaces preview while the resolution job is still in progress, and so the
+    // results will come up empty and overwrite the previous results in the Apollo Client
+    // cache while this is happening.
+    const [cachedWorkspacesPreview, setCachedWorkspacesPreview] = useState<PreviewBatchSpecWorkspaceFields[]>()
+
+    // We copy the results from `connection` to `cachedWorkspacesPreview` whenever a
+    // resolution job completes.
+    useEffect(() => {
+        if (resolutionState === BatchSpecWorkspaceResolutionState.COMPLETED && connection?.nodes.length) {
+            setCachedWorkspacesPreview(connection.nodes)
+        }
+    }, [resolutionState, connection?.nodes])
+
+    // We will instruct `WorkspacesPreviewList` to show the cached results instead of
+    // whatever is in `connection` if we know the workspaces preview resolution is
+    // currently in progress.
+    const showCached = useMemo(
+        () =>
+            Boolean(
+                cachedWorkspacesPreview?.length && (isWorkspacesPreviewInProgress || resolutionState === 'CANCELED')
+            ),
+        [cachedWorkspacesPreview, isWorkspacesPreviewInProgress, resolutionState]
+    )
 
-    const clearErrorAndPreview = useCallback(() => {
-        setIsResolvingPreview(true)
-        setResolutionError(undefined)
-        preview()
-    }, [preview])
+    const ctaButton = isWorkspacesPreviewInProgress ? (
+        <Button className="mt-3 mb-2" variant="secondary" onClick={cancel}>
+            Cancel
+        </Button>
+    ) : (
+        <Button
+            className="mt-3 mb-2"
+            variant="success"
+            disabled={!!previewDisabled}
+            data-tooltip={typeof previewDisabled === 'string' ? previewDisabled : undefined}
+            onClick={preview}
+        >
+            <SearchIcon className="icon-inline mr-1" />
+            {error ? 'Retry preview' : 'Preview workspaces'}
+        </Button>
+    )
 
-    const onFinished = useCallback(() => setIsResolvingPreview(false), [])
-    // Capture state changes when workspace resolution status changes.
-    useBatchSpecWorkspaceResolution(batchSpec, { onError: setResolutionError, onFinished })
+    const [exampleOpen, setExampleOpen] = useState(false)
+    const exampleStyle = useSpring({ height: exampleOpen ? '6.5rem' : '0rem', opacity: exampleOpen ? 1 : 0 })
+
+    const ctaInstructions = isWorkspacesPreviewInProgress ? (
+        <h4 className={styles.instruction}>Hang tight while we look for matching workspaces...</h4>
+    ) : batchSpecStale ? (
+        <h4 className={styles.instruction}>Finish editing your batch spec, then manually preview repositories.</h4>
+    ) : (
+        <>
+            <h4 className={styles.instruction}>
+                {hasPreviewed ? 'Modify your' : 'Use an'} <span className="text-monospace">on:</span> statement to
```

Comment:
`Use an` feels slightly weird to me as a wording choice here for a user with a brand new batch spec. What about:

```suggestion
                {hasPreviewed ? 'Modify your' : 'Add an'} <span className="text-monospace">on:</span> statement to
```

Reply:
Amazing, apparently Rob had the same thought -- I was about to say "I agree, but let me leave a note in Figma so Rob has a chance to veto", but the designs now also say "Add" instead of "Use"! 😄